### PR TITLE
Properly record stack traces

### PR DESCRIPTION
**Pull request by cbruegg**
I really like this library, but it has caused me some headaches related to network errors. Continuations are resumed with the Exception thrown by OkHttp, but unfortunately that `Exception` contains no indication about where the network request was started. This is because OkHttp internally uses a thread pool and thus the stack trace only includes internal OkHttp code.

This PR changes the `await` function to record the current stacktrace before enqueuing the request. In case of receiving an `Exception` from OkHttp, it is set as the `cause` of the previously recorded stack trace. 

The resulting stacktrace looks like this (note the `CallAwaitTest.kt:24`!):

```
java.io.IOException: Exception occurred while awaiting Call.
	at ru.gildor.coroutines.okhttp.CallAwaitKt.await(CallAwait.kt:18)
	at ru.gildor.coroutines.okhttp.CallAwaitKt.await$default(CallAwait.kt:17)
	at ru.gildor.coroutines.okhttp.CallAwaitTest$awaitExceptionRecorded$1.doResume(CallAwaitTest.kt:24)
	at kotlin.coroutines.experimental.jvm.internal.CoroutineImpl.resume(CoroutineImpl.kt:42)
	at kotlinx.coroutines.experimental.DispatchedKt.resumeCancellable(Dispatched.kt:224)
	at kotlinx.coroutines.experimental.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:29)
	at kotlinx.coroutines.experimental.CoroutineStart.invoke(CoroutineStart.kt:109)
	at kotlinx.coroutines.experimental.AbstractCoroutine.start(AbstractCoroutine.kt:166)
	at kotlinx.coroutines.experimental.BuildersKt__BuildersKt.runBlocking(Builders.kt:44)
	at kotlinx.coroutines.experimental.BuildersKt.runBlocking(Unknown Source)
	at ru.gildor.coroutines.okhttp.CallAwaitTestKt.testBlocking(CallAwaitTest.kt:101)
	at ru.gildor.coroutines.okhttp.CallAwaitTestKt.access$testBlocking(CallAwaitTest.kt:1)
	at ru.gildor.coroutines.okhttp.CallAwaitTest.awaitExceptionRecorded(CallAwaitTest.kt:20)
[...]
Caused by: java.io.IOException
	at ru.gildor.coroutines.okhttp.CallAwaitTest$client$1.intercept(CallAwaitTest.kt:89)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)

```
Before the change, it used to be simply:

```
Caused by: java.io.IOException
	at ru.gildor.coroutines.okhttp.CallAwaitTest$client$1.intercept(CallAwaitTest.kt:89)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall$AsyncCall.execute(RealCall.java:147)
	at okhttp3.internal.NamedRunnable.run(NamedRunnable.java:32)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

Since of course recording a stack trace causes a very minor performance cost and the original `Exception` is now only the `cause`, this behavior can be disabled using an optional parameter to `await`.

While I was already working on the library anyway, I also updated the versions of its dependencies.

### CODE REVIEW

**Comment by cbruegg** [comment]
Also, in case you're ok with this change, I'd be glad to contribute it as well for your Retrofit library.

**Comment by gildor** [comment]
Christian, thanks for your contribution, appreciate this.
I will update okhttp with some chages and Kotlin 1.3 and will than check your pull request

File: build.gradle [hunk 1]
```diff
@@ -18,10 +18,10 @@ repositories {
 }
 
 dependencies {
-    compile "org.jetbrains.kotlin:kotlin-stdlib"
-    compile "org.jetbrains.kotlinx:kotlinx-coroutines-core:0.17"
-    compile "com.squareup.okhttp3:okhttp:3.8.0"
-    testCompile 'junit:junit:4.12'
+    implementation "org.jetbrains.kotlin:kotlin-stdlib:1.2.71"
```
Comment (gildor) [inline build.gradle]: Please return `compile`. Those dependencies are public and transitive and required for librayry, they actually `api`, not `implementation`, but `api` requires java-plugin, but in this case it's not so neccessary.
You can keep it for `testCompile`

Comment (cbruegg) [reply build.gradle]: Oops, you're right, sorry for that.

**Comment by cbruegg** [comment]
> I will update okhttp with some chages and Kotlin 1.3 and will than check your pull request

Just FYI, if you update OkHttp, you'll need to change your test to include this: `.body(ResponseBody.create(null, ByteArray(0)))` as done by this PR as well.

Thanks in advance for checking my PR!

**Comment by gildor** [comment]
@cbruegg Thanks for your PR, I released version 1.0 to bintray without your changes but  I think that recording of stacktraces maybe beneficial for many users of okhttp.
My only concern, that I wouldn't enable by default and probably provide some API to enable it by default (similar to kotlinx.coroutine debug mode).  I plan to work on it today or tomorrow
Thanks again

**Comment by cbruegg** [comment]
Thanks for your feedback! That's a valid point. Although actually, I'd
recommend testing first whether this problem still exists when coroutines
debug mode is turned on. I'll try and take a look later today.

Andrey Mischenko <notifications@github.com> schrieb am So., 17. März 2019,
04:41:

**Comment by gildor** [comment]
@cbruegg I agree, this is what I want to add to README, that I still want to provide this way to debug requests, I would recommend use coroutines debug mode because it allows you to see suspend point where the exception happened this is usually enough for most cases.
But capturing full stack trace also gives additional info about stack before suspend but also you pay the price of 1 additional exception with stack trace per call, so it's also not free

**Comment by gildor** [comment]
@cbruegg I'm working on stack recording, you can check the current state on master, I also plan to migrate your tests for stack recording from this PR

Also, I added a [section about different ways of debugging](https://github.com/gildor/kotlin-coroutines-okhttp/blob/master/README.md#debugging) to README, so hope it would be helpful

I'm closing your existing PR, if you have some additions please create a new one, I plan to release it in a few days

**Pull request closed without merging**